### PR TITLE
Add sphinx/breathe documentation

### LINKED ISSUE #27: Add Doxygen/Sphinx to codebase and CI
See [Sphinx/Breathe/C++ intro](https://devblogs.microsoft.com/cppblog/clear-functional-c-documentation-with-sphinx-breathe-doxygen-cmake/).
- [x] Add doxyfile and cmake
- [ ] Add sphinx/breathe and cmake (#437)
- [x] Set up ReadTheDocs (http://celeritas.readthedocs.io/)

As the [code design](https://github.com/celeritas-project/celeritas/wiki/Code-design) components take shape as part of the permanent code base, transition portions of that document to the in-code documentation. Basically, let the Wiki page be for sketching out ideas, and let the code documentation reflect the actual state of the code.

### PR DESCRIPTION
**Pull request by sethrj**
Initial support for in-depth project documentation, including C++ API rendered through Breathe (when building locally), and integration with ReadTheDocs (see [example render](https://celeritas-sethrj.readthedocs.io/index.html); [Celeritas master documentation will be here](https://celeritas.readthedocs.io/en/latest/). Closes #27.

Breathe is definitely less stable than Doxygen so we probably want to be very selective about what ends up as documented stable API. There are warnings that can't currently be suppressed.

### CODE REVIEW

File: scripts/spack.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  concretization: together
+  packages:
+    root:
+      variants: ~aqua ~davix ~examples ~opengl ~x ~tbb ~math ~gsl
+    all:
+      providers:
+        blas: [openblas]
+        lapack: [openblas]
+        mpi: [openmpi]
+      variants: cxxstd=17
```
Comment (drbenmorgan) [inline scripts/spack.yaml:36]: Does this render the explicit `cxxstd=17` definitions for Geant4, root and vecgeom redundant?

Comment (sethrj) [reply scripts/spack.yaml:36]: They don't serve quite the same purpose: the `specs` are required and will fail if those variants don't exist. The variants under `packages` are "preferences" and are ignored if not applicable. Further complicating this process is that spack combines the two (in addition to site preferences from `packages.yaml`) in surprising ways...

Comment (drbenmorgan) [reply scripts/spack.yaml:36]: Ah gotcha, makes sense!

**Review by drbenmorgan — APPROVED** [review]

**Comment by sethrj** [comment]
Thanks for the review @drbenmorgan !

**Pull request merged**